From: Eric Moore <emoore@lsil.com>
Subject: Fixup SAS end device handling
Is-Mainline: 2.6.17

As per suggestion from hch the end device handling should rather be
done in the transport class, not the individual driver.

Signed-off-by: Hannes Reinecke <hare@suse.de>

diff -uarN b/drivers/message/fusion/mptsas.c a/drivers/message/fusion/mptsas.c
--- b/drivers/message/fusion/mptsas.c	2006-04-19 16:30:37.000000000 -0600
+++ a/drivers/message/fusion/mptsas.c	2006-04-20 09:15:37.000000000 -0600
@@ -450,9 +450,6 @@
 		for (i = 0; i < p->num_phys; i++) {
 			if (p->phy_info[i].attached.sas_address ==
 					rphy->identify.sas_address) {
-				if (!mptsas_is_end_device(
-					&p->phy_info[i].attached))
-					goto out_no_dev;
 				target_id = p->phy_info[i].attached.id;
 				vtarget->bus_id = p->phy_info[i].attached.channel;
 				vdev->lun = sdev->lun;
@@ -472,7 +469,6 @@
 			}
 		}
 	}
- out_no_dev:
 	mutex_unlock(&hd->ioc->sas_topology_mutex);
 
 	kfree(vdev);
diff -uarN b/drivers/scsi/scsi_transport_sas.c a/drivers/scsi/scsi_transport_sas.c
--- b/drivers/scsi/scsi_transport_sas.c	2006-04-12 11:21:58.000000000 -0600
+++ a/drivers/scsi/scsi_transport_sas.c	2006-04-20 09:17:18.000000000 -0600
@@ -845,6 +845,8 @@
 	    (identify->target_port_protocols &
 	     (SAS_PROTOCOL_SSP|SAS_PROTOCOL_STP|SAS_PROTOCOL_SATA)))
 		rphy->scsi_target_id = sas_host->next_target_id++;
+	else if (identify->device_type == SAS_END_DEVICE)
+		rphy->scsi_target_id = -1;
 	mutex_unlock(&sas_host->lock);
 
 	if (identify->device_type == SAS_END_DEVICE &&
